src/common/checksum.c

```diff
@@ -974,19 +978,29 @@ daos_csummer_verify_iod(struct daos_csummer *obj, daos_iod_t *iod,
 				&iod_csum->ic_data[i]);
 		if (!match) {
 			if (iod->iod_type == DAOS_IOD_ARRAY)
-				D_ERROR("Data corruption found for recx: "DF_RECX". "
-					"Calculated "DF_CI" != "
-					"received "DF_CI"\n",
-					DP_RECX(iod->iod_recxs[i]),
-					DP_CI(new_iod_csums->ic_data[i]),
-					DP_CI(iod_csum->ic_data[i]));
+				if (iod_csum->ic_data)
```

Comment:
Line 978 has already dereferenced iod_csum->ic_data so maybe this check goes above that? However, I think the only way ic_data would be null is if ic_nr == 0, which occurs when iod_size == 0 and that check is above on 965. But now that I write that out it seems somewhat convoluted so the additional check is probably a good idea.

Reply:
the reason i have this change is that the daos checksum tests did segfault here without it :-)

Reply:
Oh, interesting. That's an EC test that failed. @liuxuezhao developed most of the EC/checksum code. It would be good for him to look at these changes. 
@liuxuezhao do you know how at line checksum.c:965 iod->iod_size can be NOT 0, but still have an iod_csum->ic_data that is NULL?

Reply:
@liuxuezhao any feedback on this?

Reply:
you mean if change "if (iod_csum->ic_data)" to "if (1)", it will segfault when running test case "DAOS_EC_CSUM05"? I tested just now at that but it did not fail.
TBH I have no idea why when iod_size is non-zero, but iod_csum->ic_data is NULL, is there a test can reproduce it?

Reply:
just to be clear. all i did in this PR at first is just not call malloc if ic_nr is 0. so remove the update from this file.
then just run daos_test  -z -u CS50 and you will see the failure / segfault

the reason of the original change is to not call malloc for 0 size since 1) memory tracers complain on it 2) it is not guaranteed that the result is not NULL.

Reply:
I still cannot reproduce it by "daos_test -z -u 50", even if comment out "if (iod_csum->ic_nr)" in RPC DECODING proc func.

"not call malloc for 0 size" looks fine for me.
what I am not sure is why need "if (iod_csum->ic_data)" here

Reply:
I think the problem is why "iod_csum->ic_data" could be NULL here, and if it is NULL at L987, why it did not segfault at L978.

Reply:
i did see it when running some tests. are you sure you have enough engines when trying to run? i had 4 server / 8 engines

Reply:
right, i ran with 4 srvs, 8 engines, 8 target each engine. just run daos_test -z -u 50.

Reply:
I was able to reproduce the segfault by reverting the changes in checksum.c and running daos_test -z -u50. What I'm not understanding is why for the DAOS_EC_CSUM05 test, iod[0].iod_size == 0 in srv_obj.c:csum_add2iods. The test clearly sets the size to 195. 
This is some printing ...
```
boro-13: 08/02-18:24:59.06 boro-13 DAOS[3074149/5/220] object ERR  src/object/srv_obj.c:905 csum_add2iods() [RYON] src/object/srv_obj.c:905 [csum_add2iods()] > iods[0].iod_size: 0
boro-13: 08/02-18:24:59.06 boro-13 DAOS[3074149/5/220] object ERR  src/object/srv_obj.c:905 csum_add2iods() [RYON] src/object/srv_obj.c:905 [csum_add2iods()] > iods[1].iod_size: 5822
boro-13: 08/02-18:24:59.06 boro-13 DAOS[3074149/5/220] object ERR  src/object/srv_obj.c:905 csum_add2iods() [RYON] src/object/srv_obj.c:905 [csum_add2iods()] > iods[2].iod_size: 6162
boro-63: 08/02-18:24:59.06 boro-63 DAOS[663219/6/220] object ERR  src/object/srv_obj.c:905 csum_add2iods() [RYON] src/object/srv_obj.c:905 [csum_add2iods()] > iods[0].iod_size: 0
boro-63: 08/02-18:24:59.06 boro-63 DAOS[663219/6/220] object ERR  src/object/srv_obj.c:905 csum_add2iods() [RYON] src/object/srv_obj.c:905 [csum_add2iods()] > iods[1].iod_size: 5822
boro-63: 08/02-18:24:59.06 boro-63 DAOS[663219/6/220] object ERR  src/object/srv_obj.c:905 csum_add2iods() [RYON] src/object/srv_obj.c:905 [csum_add2iods()] > iods[2].iod_size: 6162
```

@liuxuezhao , do you have any idea why the size is 0 here? I'll also continue trying to trace it back. I'd like to understand this before we hide a potential problem.

Reply:
@ryon-jensen I think the iod_size 0 is come from test code in multiple_ec_singv_csum() L2686 it provided 0 iod_size for fetch -
2686 >       iod[0].iod_size = 0;                                                                        
2687 >       daos_fail_loc_set(DAOS_CSUM_CORRUPT_FETCH | DAOS_FAIL_ALWAYS);                              
2688 >       rc = daos_obj_fetch(oh, DAOS_TX_NONE, 0, &dkey, 3, iod, sgl, NULL, NULL);                   
2689 >       assert_rc_equal(rc, -DER_CSUM);                                                             
2690 >       assert_rc_equal(iod[0].iod_size, 195);    

In srv_obj.c, it got the real iod_size (195) by -
1492 >       >       rc = vos_fetch_begin(ioc->ioc_vos_coh, orw->orw_oid,                                
1493 >       >       >       >            orw->orw_epoch, dkey, iods_nr, iods,                           
1494 >       >       >       >            cond_flags | fetch_flags, shadows, &ioh, dth); 

but the iods above is different with "orw->orw_iod_array.oia_iods" below -
1596 >       >       >       rc = csum_add2iods(ioh, orw->orw_iod_array.oia_iods,                        
1597 >       >       >       >       >          orw->orw_iod_array.oia_iod_nr,                           
1598 >       >       >       >       >          skips, ioc->ioc_coc->sc_csummer,                         
1599 >       >       >       >       >          orwo->orw_iod_csums.ca_arrays,                           
1600 >       >       >       >       >          orw->orw_oid, &orw->orw_dkey);   

So here need the change from your PR https://github.com/daos-stack/daos/pull/12596/files#diff-675c44126216c7353850735b36e34f1145d6c16bdda01b595757666b84264f25R1586-R1595 

So back to @mchaarawi 's this PR, I think it is fine to only alloc only "if (iod_csum->ic_nr)" in crt_proc_struct_dcs_iod_csums_adv, just that in daos_csummer_compare_csum_info() seems unnecessary to check "if (iod_csum->ic_data)" again.

Reply:
OK thanks. I pushed an update to remove that. didn't test it though.
let's check the CI result in this case.

Reply:
np, there are some CI test failures, I just did a merge to include @ryon-jensen 's PR12596, to see if those failures can be fixed or not. thx

Reply:
I believe those were the same failures i saw before i added my change to the debugging statement that is trying to access that field ;-)